src/ucs/memory/rcache.c

```diff
@@ -844,14 +842,31 @@ ucs_rcache_check_overlap(ucs_rcache_t *rcache, ucs_pgt_addr_t *start,
         ucs_rcache_region_trace(rcache, region,
                                 "merge 0x%lx..0x%lx "UCS_RCACHE_PROT_FMT" with",
                                 *start, *end, UCS_RCACHE_PROT_ARG(*prot));
-        *start  = ucs_min(*start, region->super.start);
-        *end    = ucs_max(*end,   region->super.end);
-        *merged = 1;
+        *alignment = ucs_max(*alignment, region->alignment);
+        *start     = ucs_align_down_pow2(ucs_min(*start, region->super.start),
+                                         *alignment);
+        *end       = ucs_align_up_pow2(ucs_max(*end, region->super.end),
+                                       *alignment);
```

Comment:
IMO we don't need to align up/down here, but only after finished going over all overlapping regions